### PR TITLE
MAINT: add main thread check for ui dispatch, solve no ui failure

### LINKED ISSUE #1732: Unexpected failure mode for `dispatch="ui"` when there is no UI
When using `dispatch="ui"` and there is no UI (in particular, Pyface has not done toolkit selection), the dispatcher fails with an attempt to call `None` as the handler:
``` python console
>>> from traits.api import HasTraits, Int, observe
>>> class Test(HasTraits):
...    x = Int()
... 
>>> t = Test()
>>> def handler(event):
...     print(event)
... 
>>> t.observe(handler, 'x', dispatch='ui')
>>> t.x = 1 
Exception occurred in traits notification handler for event object: TraitChangeEvent(object=<__main__.Test object at 0x10de8a270>, name='x', old=0, new=1)
Traceback (most recent call last):
  File "/Users/cwebster/.edm/envs/enclosure-38/lib/python3.8/site-packages/traits/observation/_trait_event_notifier.py", line 122, in __call__
    self.dispatcher(handler, event)
  File "/Users/cwebster/.edm/envs/enclosure-38/lib/python3.8/site-packages/traits/trait_notifiers.py", line 50, in ui_dispatch
    ui_handler(handler, *args, **kw)
TypeError: 'NoneType' object is not callable
```

This also happens when the Pyface toolkit is `null` (which may be a Pyface bug):
``` python console
>>> from traits.etsconfig.api import ETSConfig
>>> ETSConfig.toolkit = 'null'
>>> import pyface.api
>>> t.x = 2     
Exception occurred in traits notification handler for event object: TraitChangeEvent(object=<__main__.Test object at 0x10de8a270>, name='x', old=1, new=2)
Traceback (most recent call last):
  File "/Users/cwebster/.edm/envs/enclosure-38/lib/python3.8/site-packages/traits/observation/_trait_event_notifier.py", line 122, in __call__
    self.dispatcher(handler, event)
  File "/Users/cwebster/.edm/envs/enclosure-38/lib/python3.8/site-packages/traits/trait_notifiers.py", line 50, in ui_dispatch
    ui_handler(handler, *args, **kw)
TypeError: 'NoneType' object is not callable
```

I'm not entirely sure what the correct behaviour is here.

I think I would expect dispatch to work if the current thread is the main thread, given the short-circuit in the code here: https://github.com/enthought/traits/blob/92ecd7e85ca399f8d4bc798d5d7071804fa5c0fb/traits/trait_notifiers.py#L46-L50

But given there is no way to dispatch something back to the main thread if there is no event loop around, I'd expect at least a better failure message if the dispatch is not on the main thread.

### PR DESCRIPTION
**Pull request by homosapien-lcy**
Previously the ui_dispatch(handler, *args, **kw) function in traits/traits/trait_notifiers.py only checks current_thread().ident with ui_thread. But supposedly, this current thread id should be compared with the main thread (threading.main_thread()). This discrepancy caused a no ui failure in ui dispatch as mentioned in issue #1732. This PR attempts to resolve this failure. Closes #1732 

**Checklist**
- [ ] Tests
- [x] Update API reference (`docs/source/traits_api_reference`) (seems no need to change for this PR)
- [ ] Update User manual (`docs/source/traits_user_manual`) (seems no need to change for this PR)
- [x] Update type annotation hints in stub files (seems no need to change for this PR)

### CODE REVIEW

**Review by mdickinson — CHANGES_REQUESTED** [review]
Like any behaviour-changing PR, this should have tests.

File: traits/trait_notifiers.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def set_ui_handler(handler):
 
 
 def ui_dispatch(handler, *args, **kw):
-    if threading.current_thread().ident == ui_thread:
+    if threading.current_thread().ident == threading.main_thread().ident:
```
Comment (mdickinson) [inline traits/trait_notifiers.py:47]: No need for the `.ident` here: you can just compare the thread objects directly.
```suggestion
    if threading.current_thread() == threading.main_thread():
```

Comment (homosapien-lcy) [reply traits/trait_notifiers.py:47]: Thanks for the suggestion. Are directly comparing threads vs comparing their ids completely equivalent?

**Comment by homosapien-lcy** [comment]
> Like any behaviour-changing PR, this should have tests.

So, for the test, I'm thinking of checking whether the handler triggered  in the main thread is a "handler" and not in the main thread is a "ui_handler". Does that sound reasonable to you? And any suggested test that I can look at for reference?

File: traits/trait_notifiers.py (hunk 1 above)
Comment (mdickinson) [reply traits/trait_notifiers.py:47]: I think so, yes. The `==` check is essentially an `is` check (there's no `__eq__` special method created), and there should be exactly one `Thread` object per OS thread.

**Comment by mdickinson** [comment]
For the test, I suggest using something like Corran's example from the linked issue. The issue is about the behaviour of `ui_dispatch`, and we should focus on testing that behaviour rather than testing implementation details. Ideally, your test should fail before this change and pass afterwards.

**Comment by homosapien-lcy** [comment]
I found that the error raised by the ui_dispatch is happened not in the main thread, thus unable to be caught in the main thread, even if it says typeError, the unittest will still past...:

```
Exception occurred in traits notification handler for event object: TraitChangeEvent(object=<traits.tests.test_trait_notifiers.TestTraitNotifiers.test_ui_dispatch.<locals>.DispatchTest object at 0x1046c58a0>, name='test_param', old=0, new=1)
Traceback (most recent call last):
  File "/Users/cyliu/Documents/3.11_test/traits/traits/observation/_trait_event_notifier.py", line 122, in __call__
    self.dispatcher(handler, event)
  File "/Users/cyliu/Documents/3.11_test/traits/traits/trait_notifiers.py", line 50, in ui_dispatch
    ui_handler(handler, *args, **kw)
TypeError: 'NoneType' object is not callable
.
----------------------------------------------------------------------
Ran 1 test in 0.002s
```

So, what I also did is add a try except in ui_dispatch to throw a systemExit that will also notify the main thread.

**Comment by homosapien-lcy** [comment]
> I've unchecked the "Update User manual" entry in the checklist: we should check whether the current documentation needs to be updated.
> 
> In particular, the documentation should make it clear what the behaviour of ui dispatch is when there's no event loop running.

Both test and documentation are added

**Comment by homosapien-lcy** [comment]
Some tests failed (for instance, Tests/tests (ubuntu-latest, 3.8)) due to unable to download certain packages:

```
1m 19s
Run sudo apt-get update
Get:[1](https://github.com/enthought/traits/actions/runs/4740125152/jobs/8415614428?pr=1740#step:3:1) https://packages.microsoft.com/ubuntu/22.04/prod jammy InRelease [3611 B]
Get:2 https://packages.microsoft.com/ubuntu/22.04/prod jammy/main amd64 Packages [60.2 kB]
Hit:3 https://ppa.launchpadcontent.net/ubuntu-toolchain-r/test/ubuntu jammy InRelease
Ign:4 http://azure.archive.ubuntu.com/ubuntu jammy InRelease
Ign:5 http://azure.archive.ubuntu.com/ubuntu jammy-updates InRelease
Ign:6 http://azure.archive.ubuntu.com/ubuntu jammy-backports InRelease
Ign:7 http://azure.archive.ubuntu.com/ubuntu jammy-security InRelease
Ign:4 http://azure.archive.ubuntu.com/ubuntu jammy InRelease
Ign:5 http://azure.archive.ubuntu.com/ubuntu jammy-updates InRelease
Ign:6 http://azure.archive.ubuntu.com/ubuntu jammy-backports InRelease
Ign:7 http://azure.archive.ubuntu.com/ubuntu jammy-security InRelease
Ign:4 http://azure.archive.ubuntu.com/ubuntu jammy InRelease
Ign:5 http://azure.archive.ubuntu.com/ubuntu jammy-updates InRelease
Ign:6 http://azure.archive.ubuntu.com/ubuntu jammy-backports InRelease
Ign:7 http://azure.archive.ubuntu.com/ubuntu jammy-security InRelease
Err:4 http://azure.archive.ubuntu.com/ubuntu jammy InRelease
  Could not connect to azure.archive.ubuntu.com:80 (52.147.219.192), connection timed out
Err:5 http://azure.archive.ubuntu.com/ubuntu jammy-updates InRelease
  Unable to connect to azure.archive.ubuntu.com:http:
Err:6 http://azure.archive.ubuntu.com/ubuntu jammy-backports InRelease
  Unable to connect to azure.archive.ubuntu.com:http:
Err:7 http://azure.archive.ubuntu.com/ubuntu jammy-security InRelease
  Unable to connect to azure.archive.ubuntu.com:http:
Fetched 63.8 kB in 37s ([17](https://github.com/enthought/traits/actions/runs/4740125152/jobs/8415614428?pr=1740#step:3:18)09 B/s)
Reading package lists...
W: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/dists/jammy/InRelease  Could not connect to azure.archive.ubuntu.com:80 (52.147.2[19](https://github.com/enthought/traits/actions/runs/4740125152/jobs/8415614428?pr=1740#step:3:20).192), connection timed out
W: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/dists/jammy-updates/InRelease  Unable to connect to azure.archive.ubuntu.com:http:
W: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/dists/jammy-backports/InRelease  Unable to connect to azure.archive.ubuntu.com:http:
W: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/dists/jammy-security/InRelease  Unable to connect to azure.archive.ubuntu.com:http:
W: Some index files failed to download. They have been ignored, or old ones used instead.
Reading package lists...
Building dependency tree...
Reading state information...
The following additional packages will be installed:
  libegl-mesa0
The following NEW packages will be installed:
  libegl-mesa0 libegl1
0 upgraded, 2 newly installed, 0 to remove and [20](https://github.com/enthought/traits/actions/runs/4740125152/jobs/8415614428?pr=1740#step:3:21) not upgraded.
Need to get 140 kB of archives.
After this operation, 465 kB of additional disk space will be used.
Ign:1 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 libegl-mesa0 amd64 22.2.5-0ubuntu0.1~22.04.1
Ign:2 http://azure.archive.ubuntu.com/ubuntu jammy/main amd64 libegl1 amd64 1.4.0-1
Ign:1 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 libegl-mesa0 amd64 22.2.5-0ubuntu0.1~22.04.1
Ign:2 http://azure.archive.ubuntu.com/ubuntu jammy/main amd64 libegl1 amd64 1.4.0-1
Ign:1 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 libegl-mesa0 amd64 22.2.5-0ubuntu0.1~22.04.1
Ign:2 http://azure.archive.ubuntu.com/ubuntu jammy/main amd64 libegl1 amd64 1.4.0-1
Err:1 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 libegl-mesa0 amd64 22.2.5-0ubuntu0.1~22.04.1
  Could not connect to azure.archive.ubuntu.com:80 (52.147.[21](https://github.com/enthought/traits/actions/runs/4740125152/jobs/8415614428?pr=1740#step:3:22)9.192), connection timed out
Err:2 http://azure.archive.ubuntu.com/ubuntu jammy/main amd64 libegl1 amd64 1.4.0-1
  Unable to connect to azure.archive.ubuntu.com:http:
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/m/mesa/libegl-mesa0_[22](https://github.com/enthought/traits/actions/runs/4740125152/jobs/8415614428?pr=1740#step:3:23).2.5-0ubuntu0.1%7e22.04.1_amd64.deb  Could not connect to azure.archive.ubuntu.com:80 (52.147.219.192), connection timed out
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/libg/libglvnd/libegl1_1.4.0-1_amd64.deb  Unable to connect to azure.archive.ubuntu.com:http:
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
Error: Process completed with exit code 100.
```

**Comment by mdickinson** [comment]
@homosapien-lcy Thanks for the updates!

For the test: this looks like a great start - the setup looks like exactly what we need. For the test itself, the behaviour we want to test is that the observer is executed when the trait value is changed. You should be able to change the observer to something that has an observable side-effect, and then test that side-effect to verify that the observer was called. (For example, set up an empty `events` list and use `events.append` as the observer.)

For the business logic: please could you explain the reason for the `try / except` and the `raise SystemExit`?

I haven't reviewed the documentation yet.

**Comment by mdickinson** [comment]
> Some tests failed [...]

This looks like an upstream problem that we've seen a few times before. I've restarted the test run.

**Comment by mdickinson** [comment]
> Some tests failed [...]

Looks like it was indeed an intermittent upstream problem - the re-run seems fine.

**Comment by mdickinson** [comment]
> For the business logic: please could you explain the reason for the `try / except` and the `raise SystemExit`?

Apologies - I see you already explained this in an earlier comment. Thank you.

It's important to think through the consequences of any change to the business logic. A question for you: with this change, if I have a Traits-using application and some part of that application happens to cause an exception to be raised in a handler, what then happens to that application? What would have happened before this change?

**Comment by homosapien-lcy** [comment]
> > For the business logic: please could you explain the reason for the `try / except` and the `raise SystemExit`?
> 
> Apologies - I see you already explained this in an earlier comment. Thank you.
> 
> It's important to think through the consequences of any change to the business logic. A question for you: with this change, if I have a Traits-using application and some part of that application happens to cause an exception to be raised in a handler, what then happens to that application? What would have happened before this change?

Thanks for the comments, I found a work around by adding an event list in the main thread and has the test handler add item to the event list, when the side thread failed, the list will be empty and cause an exception in the main thread.

File: traits/tests/test_trait_notifiers.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        def test_handler(event):
+            event_list.append(event)
+
+        # When
+        t.observe(test_handler, 'test_param', dispatch='ui')
+
+        # Then
+        try:
+            t.test_param = 1
+        except Exception:
```
Comment (mdickinson) [inline traits/tests/test_trait_notifiers.py:34]: Let's remove this try/except - it's just extra boilerplate and we don't really need it. If an exception is raised, the test will already fail, and there's no particular reason to expect an exception to be raised either before or after the change in this PR.

File: traits/tests/test_trait_notifiers.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            self.fail("test_ui_dispatch raised an Exception unexpectedly!")
+
+        # also check the observer is called (test_handler function)
+        self.assertTrue(event_list)
```
Comment (mdickinson) [inline traits/tests/test_trait_notifiers.py:38]: I'd suggest a slightly stronger check, that the event list has length one. It would be even better to extract the event from the event list and check that the "new" has the expected value of `1`.

**Comment by mdickinson** [comment]
> when the side thread failed

Just for clarity, there is no side thread here. Everything is happening in the main thread.

File: traits/tests/test_trait_notifiers.py (hunk 2 above)
Comment (mdickinson) [reply traits/tests/test_trait_notifiers.py:34]: I'd also move the `t.test_param = 1` to the "When" block - it's the main part of the behaviour we're testing. The "Then" block is for testing the effects of that behaviour.

Comment (mdickinson) [reply traits/tests/test_trait_notifiers.py:34]: It's often helpful to think about expanding out the "When/Then" to full sentence fragments: E.g., "**When** we change the value of a trait with a `dispatch='ui'` observer, *then* that observer is run as expected.

File: traits/tests/test_trait_notifiers.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+import unittest
+from traits.api import HasTraits, Int
```
Comment (mdickinson) [inline traits/tests/test_trait_notifiers.py:12]: Please follow the usual convention for import layout - there should be a blank line between import groups. `import unittest` belongs to the standard library import group, while the `from traits` import belongs to the 1st party imports.

**Comment by mdickinson** [comment]
> Just for clarity, there is no side thread here. Everything is happening in the main thread.

Actually, this is a good point. It would be useful to have tests that check that we get the expected result when `dispatch=ui` is used from a background thread, too. That's independent of solving this particular issue, though - we can leave it for this PR and do it in a separate PR.

File: traits/tests/test_trait_notifiers.py (hunk 2 above)
Comment (homosapien-lcy) [reply traits/tests/test_trait_notifiers.py:34]: Thanks! thats a great advice

**Comment by homosapien-lcy** [comment]
> I found that the error raised by the ui_dispatch is happened not in the main thread, thus unable to be caught in the main thread, even if it says typeError, the unittest will still past...:
> 
> ```
> (py311) (base) cyliu@aus552cyliu traits % python3.11 -m unittest traits/tests/test_trait_notifiers.py
> Exception occurred in traits notification handler for event object: TraitChangeEvent(object=<traits.tests.test_trait_notifiers.TestTraitNotifiers.test_ui_dispatch.<locals>.DispatchTest object at 0x10929e840>, name='test_param', old=0, new=1)
> Traceback (most recent call last):
>   File "/Users/cyliu/Documents/3.11_test/traits/traits/observation/_trait_event_notifier.py", line 122, in __call__
>     self.dispatcher(handler, event)
>   File "/Users/cyliu/Documents/3.11_test/traits/traits/trait_notifiers.py", line 50, in ui_dispatch
>     ui_handler(handler, *args, **kw)
> TypeError: 'NoneType' object is not callable
> .
> ----------------------------------------------------------------------
> Ran 1 test in 0.004s
> 
> OK
> ```
> 
> So, what I also did is add a try except in ui_dispatch to throw a systemExit that will also notify the main thread.

Thanks for the explanation, but I have a question then: if everything is happening in the main thread, why the "TypeError" here does not cause the test to fail? The original test looks like this:

```
class TestTraitNotifiers(unittest.TestCase):
    def test_ui_dispatch(self):
        class DispatchTest(HasTraits):
            test_param = Int()
        t = DispatchTest()

        event_list = []

        # create handler
        def test_handler(event):
            event_list.append(event)

        t.observe(test_handler, 'test_param', dispatch='ui')
        try:
            t.test_param = 1
        except:
            self.fail("Test failed")
```

File: traits/tests/test_trait_notifiers.py (hunk 3 above)
Comment (homosapien-lcy) [reply traits/tests/test_trait_notifiers.py:35]: Stronger tests are added

**Comment by mdickinson** [comment]
> if everything is happening in the main thread, why the "TypeError" here does not cause the test to fail?

By default, the exception is caught and logged. What you're seeing here in the output is the logged message. See the code here: https://github.com/enthought/traits/blob/dcbefb4e0a94a352a199ca92b3216163b79c5366/traits/observation/_trait_event_notifier.py#L118-L124

and here: https://github.com/enthought/traits/blob/dcbefb4e0a94a352a199ca92b3216163b79c5366/traits/observation/exception_handling.py#L37-L49

In cases like these, a useful technique is to search for the error message in the codebase - a code search for "Exception occurred in traits notification handler" would have brought up the above location.

You can easily verify that everything is happening in the main thread by adding a `print(threading.current_thread())` debugging call.

File: traits/tests/test_trait_notifiers.py (hunk 3 above)
Comment (mdickinson) [reply traits/tests/test_trait_notifiers.py:35]: Thanks. Please could you check that the event list has length one, too? e.g.,, replace `self.assertTrue(event_list)` with `self.assertEqual(len(event_list), 1)`?

**Comment by homosapien-lcy** [comment]
> > if everything is happening in the main thread, why the "TypeError" here does not cause the test to fail?
> 
> By default, the exception is caught and logged. What you're seeing here in the output is the logged message. See the code here:
> 
> https://github.com/enthought/traits/blob/dcbefb4e0a94a352a199ca92b3216163b79c5366/traits/observation/_trait_event_notifier.py#L118-L124
> 
> and here:
> 
> https://github.com/enthought/traits/blob/dcbefb4e0a94a352a199ca92b3216163b79c5366/traits/observation/exception_handling.py#L37-L49
> 
> In cases like these, a useful technique is to search for the error message in the codebase - a code search for "Exception occurred in traits notification handler" would have brought up the above location.
> 
> You can easily verify that everything is happening in the main thread by adding a `print(threading.current_thread())` debugging call.

So, the exception already happened in the main thread but it is caught thus the test is not failed. Is that what you mean?

**Comment by mdickinson** [comment]
> If I under stand you correctly: The exception already happened in the main thread but it is caught thus the test is not failed?

Yes! But you don't need to take my word for it - you can verify this for yourself.

> And your comments are addressed

Please could you address [this one](https://github.com/enthought/traits/pull/1740/files#r1173395668)? https://github.com/enthought/traits/pull/1740/files#r1173395668

File: docs/source/traits_user_manual/notification.rst [hunk 5]
```diff
@@ -617,6 +617,23 @@ used for putting up the observer::
     instance.observe(handler, "name", dispatch="ui")
     instance.observe(handler, "name", dispatch="ui", remove=True)
 
+If dispatch is set to "ui", the ui_dispatch(handler, ..), the behavior
+of the function will be different depending on whether the current thread is the
+main thread.
```
Comment (mdickinson) [inline docs/source/traits_user_manual/notification.rst:622]:
```suggestion
main thread or not.
```

**Comment by mdickinson** [comment]
@homosapien-lcy Please ignore my comments on the documentation. There are a number of changes needed there, but I think the easiest path is for me to make a commit to this branch with the necessary changes.

File: traits/tests/test_trait_notifiers.py (hunk 3 above)
Comment (homosapien-lcy) [reply traits/tests/test_trait_notifiers.py]: Thanks for the suggestion, just updated

**Comment by mdickinson** [comment]
I've added new documentation, and removed the documentation that was added to the migration section.

**Review by mdickinson — APPROVED** [review]
I think everything has now been addressed. I'll merge when the CI completes.

@homosapien-lcy Thank you for the fix.
